diff --git a/lld/wasm/Writer.cpp b/lld/wasm/Writer.cpp
index 2bf4b370a7db..3c1bd3bb2250 100644
--- a/lld/wasm/Writer.cpp
+++ b/lld/wasm/Writer.cpp
@@ -1173,7 +1173,8 @@ void Writer::createSyntheticInitFunctions() {
     };
     if (llvm::any_of(segments, hasTLSRelocs)) {
       ctx.sym.applyTLSRelocs = symtab->addSyntheticFunction(
-          "__wasm_apply_tls_relocs", WASM_SYMBOL_VISIBILITY_HIDDEN,
+          "__wasm_apply_tls_relocs",
+          WASM_SYMBOL_VISIBILITY_DEFAULT | WASM_SYMBOL_EXPORTED,
           make<SyntheticFunction>(nullSignature, "__wasm_apply_tls_relocs"));
       ctx.sym.applyTLSRelocs->markLive();
     }
@@ -1363,6 +1364,14 @@ void Writer::createInitMemoryFunction() {
           writeUleb128(os, s->index, "segment index immediate");
           writeU8(os, 0, "memory index immediate");
         }
+
+        // After initializing the TLS segment, we also need to apply TLS
+        // relocations in the same way __wasm_init_tls does.
+        if (s->isTLS() && ctx.sym.applyGlobalTLSRelocs) {
+          writeU8(os, WASM_OPCODE_CALL, "CALL");
+          writeUleb128(os, ctx.sym.applyGlobalTLSRelocs->getFunctionIndex(),
+                      "function index");
+        }
       }
     }
 
@@ -1515,8 +1524,8 @@ void Writer::createApplyGlobalRelocationsFunction() {
 }
 
 // Similar to createApplyGlobalRelocationsFunction but for
-// TLS symbols.  This cannot be run during the start function
-// but must be delayed until __wasm_init_tls is called.
+// TLS symbols. Can only be called after __tls_base is
+// initialized.
 void Writer::createApplyGlobalTLSRelocationsFunction() {
   // First write the body's contents to a string.
   std::string bodyContent;
diff --git a/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp b/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp
index 02db1b142a22..2b26bf778262 100644
--- a/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp
+++ b/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp
@@ -1881,9 +1881,7 @@ WebAssemblyTargetLowering::LowerGlobalTLSAddress(SDValue Op,
   // Currently only Emscripten supports dynamic linking with threads. Therefore,
   // on other targets, if we have thread-local storage, only the local-exec
   // model is possible.
-  auto model = Subtarget->getTargetTriple().isOSEmscripten()
-                   ? GV->getThreadLocalMode()
-                   : GlobalValue::LocalExecTLSModel;
+  auto model = GV->getThreadLocalMode();
 
   // Unsupported TLS modes
   assert(model != GlobalValue::NotThreadLocal);
diff --git a/llvm/lib/Target/WebAssembly/WebAssemblyRegStackify.cpp b/llvm/lib/Target/WebAssembly/WebAssemblyRegStackify.cpp
index bc91c6424b63e..7f2eb7d9b7fd8 100644
--- a/llvm/lib/Target/WebAssembly/WebAssemblyRegStackify.cpp
+++ b/llvm/lib/Target/WebAssembly/WebAssemblyRegStackify.cpp
@@ -339,38 +339,16 @@ static bool isSafeToMove(const MachineOperand *Def, const MachineOperand *Use,
   assert(DefI->getParent() == Insert->getParent());
   assert(UseI->getParent() == Insert->getParent());
 
-  // The first def of a multivalue instruction can be stackified by moving,
-  // since the later defs can always be placed into locals if necessary. Later
-  // defs can only be stackified if all previous defs are already stackified
-  // since ExplicitLocals will not know how to place a def in a local if a
-  // subsequent def is stackified. But only one def can be stackified by moving
-  // the instruction, so it must be the first one.
-  //
-  // TODO: This could be loosened to be the first *live* def, but care would
-  // have to be taken to ensure the drops of the initial dead defs can be
-  // placed. This would require checking that no previous defs are used in the
-  // same instruction as subsequent defs.
-  if (Def != DefI->defs().begin())
+  // For now avoid stackifying any multi-def instructions. While it's
+  // theoretically possible to do so for the first def in some cases this has
+  // historically led to bugs such as #199910 and #98323. For now this
+  // conservatively skips all multi-def instructions as a consequence. Note that
+  // multi-def instructions are expected to be not all that common so this in
+  // theory doesn't have a massive impact, but nevertheless this'd still be
+  // something to optimize better in the future.
+  if (DefI->getNumExplicitDefs() > 1)
     return false;
 
-  // If any subsequent def is used prior to the current value by the same
-  // instruction in which the current value is used, we cannot
-  // stackify. Stackifying in this case would require that def moving below the
-  // current def in the stack, which cannot be achieved, even with locals.
-  // Also ensure we don't sink the def past any other prior uses.
-  for (const auto &SubsequentDef : drop_begin(DefI->defs())) {
-    auto I = std::next(MachineBasicBlock::const_iterator(DefI));
-    auto E = std::next(MachineBasicBlock::const_iterator(UseI));
-    for (; I != E; ++I) {
-      for (const auto &PriorUse : I->uses()) {
-        if (&PriorUse == Use)
-          break;
-        if (PriorUse.isReg() && SubsequentDef.getReg() == PriorUse.getReg())
-          return false;
-      }
-    }
-  }
-
   // If moving is a semantic nop, it is always allowed
   const MachineBasicBlock *MBB = DefI->getParent();
   auto NextI = std::next(MachineBasicBlock::const_iterator(DefI));
